QR_extraction_service.ipynb

```diff
@@ -112,53 +128,95 @@
         "    return observation\n",
         "\n",
         "def main():\n",
-        "    with open('qr_example.json', 'r') as f:\n",
-        "        data = json.load(f)\n",
+        "    # Extract units from questionnaire file\n",
+        "    units_map = extract_units('ht_wt_panel-q.json')\n",
         "\n",
-        "    observations = []\n",
+        "    # Parse questionnaire response JSON file\n",
+        "    with open('ht_wt_panel-qr.json', 'r') as f:\n",
+        "        questionnaire_response = json.load(f)\n",
         "\n",
-        "    # Iterate through each item\n",
-        "    for item in data['item']:\n",
+        "    # Iterate through each item in the questionnaire response\n",
+        "    for item in questionnaire_response['item']:\n",
         "        link_id = item['linkId']\n",
         "        text = item['text']\n",
-        "        answer = item['answer'][0]\n",
+        "        answer = item.get('answer', [{}])[0]\n",
         "\n",
-        "        # Create Observation resource for the item\n",
-        "        observation = create_observation(link_id, text, answer)\n",
-        "        observations.append(observation)\n",
+        "        # Get units for the current question\n",
+        "        units = units_map.get(link_id)\n",
         "\n",
-        "    # Serialize Observation resources into JSON format\n",
-        "    for index, observation in enumerate(observations):\n",
-        "        # Convert Decimal values to floats before serialization\n",
-        "        observation_dict = observation.dict()\n",
-        "        for key, value in observation_dict.items():\n",
-        "            if isinstance(value, dict):\n",
-        "                for k, v in value.items():\n",
-        "                    if isinstance(v, Decimal):\n",
-        "                        value[k] = float(v)\n",
-        "            elif isinstance(value, Decimal):\n",
-        "                observation_dict[key] = float(value)\n",
+        "        # Create Observation resource for the item\n",
+        "        observation = create_observation(link_id, text, answer, units)\n",
```

Comment:
@NRBaconPharmD I think we'll want to pass the extracted patient resources into here so you can grab the patient.id and insert it into your observation.subject field as a reference (e.g. observation.subject = patient/{[id}). I'll add my code to your main so our code is more integrated sometime tomorrow.